app/bot/__init__.py

```diff
@@ -11,7 +11,35 @@
 _intents.message_content = True
 
 client = discord.Client(intents=_intents)
+conversation_history = []
 
+# FIXME: This will probably have race condition issues, GL HF :)
+def update_conversation_history(user_message, bot_response):
+    # Append the new user message and bot response to the conversation history
+    conversation_history.append({"role": "user", "content": user_message})
+    conversation_history.append({"role": "bot", "content": bot_response})
+
+    # Optionally, limit the history length to the last N messages
+    if len(conversation_history) > 20:
+        conversation_history = conversation_history[-20:]
+
+def context_from_conversation_history():
+    context = ""
+    for message in conversation_history:
+        context += f"{message['role']}: {message['content']}\n"
+    return context
+
+def generate_response_with_context(message):
+    context = context_from_conversation_history()
+
+    response = await llm.generate_response(
+        _botconf.botconfig.system_prompt +
+        f" The user's name is {message.author.name}" +
+        context +
+        f"user: {message}\nbot:"
```

Comment:
I think this _might_ work. But the best way to do it would be using [Ollama's chat API](https://github.com/ollama/ollama/blob/main/docs/api.md#generate-a-chat-completion)

Reply:
It's not clear to me how to make that refactor. Let's try this for now, and you can poke around with the proper implementation to see if it's better.